### ISSUE TITLE
Gradualizer complains about valid subset (adding explicit type annotations / hints)

### ISSUE THREAD

**Issue by eksperimental** [esl/gradient#82]
I'm runnning on this project/commit
https://github.com/eksperimental/checker_basic/tree/bd2bcbe717863c4a1c0c7acc1866072815608e52

The code in question:

```elixir
  @spec uniq_id() :: pos_integer()
  def uniq_id() do
    System.unique_integer([:positive, :monotonic])
  end
```

Will give me this error:

```
lib/checker/util.ex: The function call on line 34 is expected to have type pos_integer() but it has type integer()
32   @spec uniq_id() :: pos_integer()
33   def uniq_id() do
34     System.unique_integer([:positive, :monotonic])
35   end
```

I know the return will be a `pos_integer()` because of the options passed to `System.unique_integer/2`.

**Comment by erszcz** [comment]
Hey, @eksperimental!

This is a valid warning, as `pos_integer()` is a subtype of `integer()`, but not the other way around. That being said, there's a way to specify the return type and make the type checker happy with https://github.com/esl/gradient/blob/370c895927263efb2ecd08ae6ca280ac58351b40/lib/gradient/type_annotation.ex#L5

The type checker cannot infer that by itself, as as the choice of the return value is based on the params at runtime, whereas the type checker reasons about the code at compile time. You know the return value will be a `pos_integer()` because of the options, but the type checker doesn't unless you use `assert_type` to tell it.

**Comment by erszcz** [comment]
So, it would be something like:

```elixir
use Gradient.TypeAnnotation

...

@spec uniq_id() :: pos_integer()
def uniq_id() do
  System.unique_integer([:positive, :monotonic]) |> assert_type(pos_integer())
end
```

**Comment by eksperimental** [comment]
Thank you for the solution. I don't know if affecting the code to please the type checker will work.
I was looking for solutions, and I know Gradualizer does not support ignore attributes yet, like Dialyzer does with `@dialyzer`.

Is there other way to ignore certain warnings?

**Comment by eksperimental** [comment]
> ```elixir
>   System.unique_integer([:positive, :monotonic]) |> assert_type(pos_integer())
> ```

It worked by calling `|> assert_type({:pos_integer, 0})`  though.

**Comment by erszcz** [comment]
Hmm, indeed, the assertion as I proposed it fails. Diff:

```
diff --git a/lib/checker/util.ex b/lib/checker/util.ex
index 2e2b030..352fa76 100644
--- a/lib/checker/util.ex
+++ b/lib/checker/util.ex
@@ -1,6 +1,8 @@
 defmodule Checker.Util do
   @moduledoc false
 
+  use Gradient.TypeAnnotation
+
   @spec fetch_url_status(Checker.url()) ::
           {:ok, Checker.http_response_status() | :unreachable} | {:error, Exception.t()}
   def fetch_url_status(url) when is_binary(url) do
@@ -30,7 +32,7 @@ defmodule Checker.Util do
 
   @spec uniq_id() :: pos_integer()
   def uniq_id() do
-    System.unique_integer([:positive, :monotonic])
+    System.unique_integer([:positive, :monotonic]) |> assert_type(pos_integer())
   end
 
   @spec pid(term) :: pid | nil
diff --git a/mix.lock b/mix.lock
index 2403eb3..60434d5 100644
--- a/mix.lock
+++ b/mix.lock
@@ -8,8 +8,8 @@
   "ex_doc": {:hex, :ex_doc, "0.28.4", "001a0ea6beac2f810f1abc3dbf4b123e9593eaa5f00dd13ded024eae7c523298", [:mix], [{:earmark_parser, "~> 1.4.19", [hex: :earmark_parser, repo: "hexpm", optional: false]}, {:makeup_elixir, "~> 0.14", [hex: :makeup_elixir, repo: "hexpm", optional: false]}, {:makeup_erlang, "~> 0.1", [hex: :makeup_erlang, repo: "hexpm", optional: false]}], "hexpm", "bf85d003dd34911d89c8ddb8bda1a958af3471a274a4c2150a9c01c78ac3f8ed"},
   "file_system": {:hex, :file_system, "0.2.10", "fb082005a9cd1711c05b5248710f8826b02d7d1784e7c3451f9c1231d4fc162d", [:mix], [], "hexpm", "41195edbfb562a593726eda3b3e8b103a309b733ad25f3d642ba49696bf715dc"},
   "finch": {:hex, :finch, "0.11.0", "622d31c224c801444c6003544fa1964a29551895b45c3543f6e4d8b0dab80f71", [:mix], [{:castore, "~> 0.1", [hex: :castore, repo: "hexpm", optional: false]}, {:mint, "~> 1.3", [hex: :mint, repo: "hexpm", optional: false]}, {:nimble_options, "~> 0.4.0", [hex: :nimble_options, repo: "hexpm", optional: false]}, {:nimble_pool, "~> 0.2.6", [hex: :nimble_pool, repo: "hexpm", optional: false]}, {:telemetry, "~> 0.4 or ~> 1.0", [hex: :telemetry, repo: "hexpm", optional: false]}], "hexpm", "047550448148d828ddea1fb76fa3f0d2d0742eca6919e548fdd8357630cae449"},
-  "gradient": {:git, "https://github.com/esl/gradient.git", "b4c2682314014ee750feda942b42da543a0ddda2", []},
-  "gradualizer": {:git, "https://github.com/josefs/Gradualizer.git", "9e629ade733780113973fe672a51d9650ed0cd86", [ref: "9e629ad"]},
+  "gradient": {:git, "https://github.com/esl/gradient.git", "6ffbff3aa62c71f282532e225354d0f290539691", []},
+  "gradualizer": {:git, "https://github.com/josefs/Gradualizer.git", "a72e7f0b7e55f235fdaa944340a13e0e41bc5ee3", [ref: "a72e7f0"]},
   "hpax": {:hex, :hpax, "0.1.1", "2396c313683ada39e98c20a75a82911592b47e5c24391363343bde74f82396ca", [:mix], [], "hexpm", "0ae7d5a0b04a8a60caf7a39fcf3ec476f35cc2cc16c05abea730d3ce6ac6c826"},
   "jason": {:hex, :jason, "1.3.0", "fa6b82a934feb176263ad2df0dbd91bf633d4a46ebfdffea0c8ae82953714946", [:mix], [{:decimal, "~> 1.0 or ~> 2.0", [hex: :decimal, repo: "hexpm", optional: true]}], "hexpm", "53fc1f51255390e0ec7e50f9cb41e751c260d065dcba2bf0d08dc51a4002c2ac"},
   "makeup": {:hex, :makeup, "1.1.0", "6b67c8bc2882a6b6a445859952a602afc1a41c2e08379ca057c0f525366fc3ca", [:mix], [{:nimble_parsec, "~> 1.2.2 or ~> 1.3", [hex: :nimble_parsec, repo: "hexpm", optional: false]}], "hexpm", "0a45ed501f4a8897f580eabf99a2e5234ea3e75a4373c8a52824f6e873be57a6"},

**Comment by eksperimental** [comment]
the main branch of that repo with current Gradient, works without problems. When i run `mix gradient` i get

> No problems found!

When I run your PR version: https://github.com/eksperimental/checker/tree/test_gradient_pull_101
I get a lot of warnings,

```
Typechecking files...
lib/checker/instance.ex: Call to undefined function Checker.Util.via/2 on line 21
lib/checker/job.ex: Call to undefined function Checker.Util.via/1 on line 14
lib/checker/job_supervisor.ex: Call to undefined function Checker.Util.via/1 on line 11
lib/checker/mock.ex: Call to undefined function Checker.Util.fetch_url_status/1 on line 80
lib/checker/server.ex: Call to undefined function Checker.Util.get_instance/1 on line 45
lib/checker/server.ex: Call to undefined function Checker.Util.get_instance/1 on line 102
lib/checker/server.ex: Call to undefined function Checker.Util.'require_keys!'/2 on line 155
lib/checker/server.ex: Call to undefined function Checker.Util.via/2 on line 20
lib/checker.ex: Call to undefined function Checker.Util.pid/1 on line 38
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 50
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 56
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 62
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 68
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 74
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 80
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 87
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 97
lib/checker.ex: Call to undefined function Checker.Util.via/1 on line 104
```

and no success message

**Comment by eksperimental** [comment]
@erszcz nevermind the previous message. It does work fine!

**Comment by erszcz** [comment]
Fixed with #101 being merged!
